File: ralph/loop.py
"""Ralph Wiggum loop - continuous retry with context reset."""

import asyncio
from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from pathlib import Path
from typing import Optional, Callable, Awaitable, List
import logging

from plan.models import Task, TaskStatus, ProjectPlan
from personas.models import AgentInstance, PersonaConfig, LifecycleState
from personas.generator import ClaudeMdGenerator
from spawner.terminal import TerminalManager, AgentProcess
from spawner.branch_manager import BranchManager
from spawner.sandbox import SandboxManager
from ci.interface import CIProvider, CIEvent
from ci.events import CIEventType

from .context import ContextSnapshot, ContextBuilder

logger = logging.getLogger(__name__)


class LoopResultType(Enum):
    """Result type from a Ralph Wiggum loop iteration."""
    TASK_COMPLETE = "task_complete"
    BLOCKED = "blocked"
    PR_CREATED = "pr_created"
    MAX_RETRIES = "max_retries"
    ERROR = "error"
    SHUTDOWN = "shutdown"


@dataclass
class LoopResult:
    """Result of a loop iteration."""
    result_type: LoopResultType
    task_id: Optional[str] = None
    summary: str = ""
    blocked_on: List[str] = None
    pr_url: Optional[str] = None
    error: Optional[str] = None
    spawn_count: int = 0
    retry_count: int = 0

    def __post_init__(self):
        if self.blocked_on is None:
            self.blocked_on = []


class RalphWiggumLoop:
    """
    The Ralph Wiggum loop - continuous retry mechanism with context reset.

    Named after the lovably persistent Ralph Wiggum from The Simpsons,
    this approach embraces iterative improvement over single-shot perfection.

    Key features:
    - Agents work until they hit a natural breakpoint
    - At breakpoints (task complete, blocked, PR created), agents are terminated
    - Fresh agents are spawned with a context summary, preventing context rot
    - CI events trigger immediate retry checks for blocked agents
    """

    def __init__(
        self,
        repo_dir: Path,
        plan: ProjectPlan,
        ci_provider: CIProvider,
        terminal_manager: TerminalManager,
        branch_manager: BranchManager,
        sandbox_manager: SandboxManager,
        comm_file_path: Path,
        max_retries: int = 100,
        retry_interval: float = 30.0,
        snapshot_dir: Optional[Path] = None,
    ):
        self.repo_dir = Path(repo_dir)
        self.plan = plan
        self.ci_provider = ci_provider
        self.terminal_manager = terminal_manager
        self.branch_manager = branch_manager
        self.sandbox_manager = sandbox_manager
        self.comm_file_path = Path(comm_file_path)
        self.max_retries = max_retries
        self.retry_interval = retry_interval
        self.snapshot_dir = Path(snapshot_dir) if snapshot_dir else repo_dir / "running" / "snapshots"

        self.context_builder = ContextBuilder(repo_dir)
        self.claude_md_generator = ClaudeMdGenerator(plan)

        self._running = False
        self._ci_event_queue: asyncio.Queue[CIEvent] = asyncio.Queue()

    async def run_agent_loop(
        self,
        agent: AgentInstance,
        task: Task,
    ) -> LoopResult:
        """
        Run the Ralph Wiggum loop for an agent.

        This method keeps spawning fresh agents until:
        - All tasks are complete
        - Max retries exceeded
        - Shutdown requested
        - Unrecoverable error

        Args:
            agent: The agent instance to run
            task: The initial task

        Returns:
            LoopResult with the final state
        """
        self._running = True
        current_task = task

        # Subscribe to CI events for this agent
        await self.ci_provider.subscribe(self._on_ci_event)

        try:
            while self._running and agent.retry_count < self.max_retries:
                # 1. Capture snapshot (for context summary)
                snapshot = self.context_builder.capture_snapshot(
                    agent, current_task, self.comm_file_path
                )
                self.context_builder.save_snapshot(snapshot, self.snapshot_dir)

                # 2. Build context summary for fresh agent
                context_summary = self.context_builder.build_context_summary(snapshot)

                # 3. Spawn fresh agent with context
                agent_process = await self._spawn_fresh_agent(
                    agent, current_task, context_summary
                )
                agent.increment_spawn()

                # 4. Wait for breakpoint
                result = await self._wait_for_breakpoint(agent, agent_process)

                # 5. Handle breakpoint
                if result.result_type == LoopResultType.TASK_COMPLETE:
                    # Task done - get next task or finish
                    logger.info(f"Agent {agent.agent_id} completed task {result.task_id}")

                    next_task = self._get_next_task(agent)
                    if next_task:
                        current_task = next_task
                        agent.set_working(current_task)
                        continue
                    else:
                        # No more tasks
                        agent.lifecycle_state = LifecycleState.COMPLETE
                        return result

                elif result.result_type == LoopResultType.BLOCKED:
                    # Blocked - terminate and wait for unblock
                    logger.info(f"Agent {agent.agent_id} blocked on {result.blocked_on}")
                    agent.set_blocked(result.blocked_on, result.summary)

                    # Terminate agent
                    self.terminal_manager.terminate(agent.agent_id)

                    # Wait for unblock signal
                    unblocked = await self._wait_for_unblock(agent, result.blocked_on)
                    if not unblocked:
                        # Timeout or shutdown
                        return result

                    # Continue loop with fresh spawn
                    agent.lifecycle_state = LifecycleState.IDLE
                    continue

                elif result.result_type == LoopResultType.PR_CREATED:
                    # PR created - terminate and wait for merge
                    logger.info(f"Agent {agent.agent_id} created PR: {result.pr_url}")
                    agent.set_pr_pending(result.pr_url)

                    # Terminate agent
                    self.terminal_manager.terminate(agent.agent_id)

                    # Wait for PR merge
                    merged = await self._wait_for_pr_merge(result.pr_url)
                    if not merged:
                        return result

                    # Continue to next task
                    next_task = self._get_next_task(agent)
                    if next_task:
                        current_task = next_task
                        agent.set_working(current_task)
                        continue
                    else:
                        agent.lifecycle_state = LifecycleState.COMPLETE
                        return result

                elif result.result_type == LoopResultType.ERROR:
                    # Error - increment retry and try again
                    logger.error(f"Agent {agent.agent_id} error: {result.error}")
                    agent.increment_retry()
                    self.terminal_manager.terminate(agent.agent_id)
                    continue

                else:
                    # Shutdown or unknown
                    return result

            # Max retries exceeded
            return LoopResult(
                result_type=LoopResultType.MAX_RETRIES,
                task_id=current_task.id if current_task else None,
                spawn_count=agent.spawn_count,
                retry_count=agent.retry_count,
            )

        finally:
            await self.ci_provider.unsubscribe(self._on_ci_event)
            self.terminal_manager.terminate(agent.agent_id)

    async def _spawn_fresh_agent(
        self,
        agent: AgentInstance,
        task: Task,
        context_summary: str,
    ) -> AgentProcess:
        """Spawn a fresh agent with context summary."""
        # Generate .claude.md with context
        claude_md_content = self.claude_md_generator.generate(
            agent.persona_config,
            task,
            agent.branch,
            context_summary,
        )

        # Inject into sandbox
        sandbox_path = self.sandbox_manager.get_sandbox(agent.agent_id)
        if not sandbox_path:
            sandbox_path = self.sandbox_manager.create_sandbox(agent.agent_id)

        self.sandbox_manager.inject_claude_md(agent.agent_id, claude_md_content)

        # Build initial prompt
        prompt = self._build_agent_prompt(agent, task)

        # Spawn the agent
        return self.terminal_manager.spawn_claude_agent(
            agent_id=agent.agent_id,
            prompt=prompt,
            working_dir=sandbox_path,
        )

    def _build_agent_prompt(self, agent: AgentInstance, task: Task) -> str:
        """Build the initial prompt for a Claude agent."""
        return f"""You are agent "{agent.agent_id}" ({agent.role}).

Your current task is: {task.id} - {task.description}

IMPORTANT: Read your .claude.md file for full context and instructions.

When you reach a natural stopping point (task complete, blocked, or PR created),
update communications.json with your breakpoint status:

For task complete:
  lifecycle_state: "complete"
  breakpoint: {{"type": "task_complete", "task_id": "{task.id}", "summary": "..."}}

For blocked:
  lifecycle_state: "blocked"
  breakpoint: {{"type": "blocked", "blocked_on": ["T001"], "reason": "..."}}

For PR created:
  lifecycle_state: "pr_pending"
  breakpoint: {{"type": "pr_created", "pr_url": "..."}}

Start by reading .claude.md and the current state of communications.json.
"""

    async def _wait_for_breakpoint(
        self,
        agent: AgentInstance,
        process: AgentProcess,
    ) -> LoopResult:
        """Wait for an agent to reach a breakpoint."""
        check_interval = 2.0  # Check every 2 seconds

        while self._running and process.is_running:
            # Check communications.json for breakpoint
            breakpoint = self._check_for_breakpoint(agent.agent_id)
            if breakpoint:
                return self._breakpoint_to_result(breakpoint)

            await asyncio.sleep(check_interval)

        # Process ended without breakpoint
        if not self._running:
            return LoopResult(result_type=LoopResultType.SHUTDOWN)

        # Check one more time for breakpoint
        breakpoint = self._check_for_breakpoint(agent.agent_id)
        if breakpoint:
            return self._breakpoint_to_result(breakpoint)

        # Process ended without proper breakpoint - treat as error
        return LoopResult(
            result_type=LoopResultType.ERROR,
            error="Agent process ended without signaling breakpoint",
        )

    def _check_for_breakpoint(self, agent_id: str) -> Optional[dict]:
        """Check communications.json for a breakpoint from an agent."""
        try:
            import json
            data = json.loads(self.comm_file_path.read_text())
            agent_data = data.get(agent_id, {})

            lifecycle_state = agent_data.get("lifecycle_state", "")
            breakpoint = agent_data.get("breakpoint")

            if lifecycle_state in ("complete", "blocked", "pr_pending") and breakpoint:
                return breakpoint

            return None
        except (json.JSONDecodeError, FileNotFoundError):
            return None

    def _breakpoint_to_result(self, breakpoint: dict) -> LoopResult:
        """Convert a breakpoint dict to a LoopResult."""
        bp_type = breakpoint.get("type", "")

        if bp_type == "task_complete":
            return LoopResult(
                result_type=LoopResultType.TASK_COMPLETE,
                task_id=breakpoint.get("task_id"),
                summary=breakpoint.get("summary", ""),
            )
        elif bp_type == "blocked":
            return LoopResult(
                result_type=LoopResultType.BLOCKED,
                blocked_on=breakpoint.get("blocked_on", []),
                summary=breakpoint.get("reason", ""),
            )
        elif bp_type == "pr_created":
            return LoopResult(
                result_type=LoopResultType.PR_CREATED,
                pr_url=breakpoint.get("pr_url"),
                task_id=breakpoint.get("task_id"),
            )
        else:
            return LoopResult(
                result_type=LoopResultType.ERROR,
                error=f"Unknown breakpoint type: {bp_type}",
            )

    async def _wait_for_unblock(
        self,
        agent: AgentInstance,
        blocked_on: List[str],
    ) -> bool:
        """Wait for blocking dependencies to be resolved."""
        while self._running and agent.retry_count < self.max_retries:
            # Check if all blockers are resolved
            if self._are_blockers_resolved(blocked_on):
                return True

            # Wait for CI event or timeout
            try:
                event = await asyncio.wait_for(
                    self._ci_event_queue.get(),
                    timeout=self.retry_interval,
                )

                # Check if this event resolves our blockers
                if self._event_resolves_blockers(event, blocked_on):
                    return True

            except asyncio.TimeoutError:
                # Periodic check
                agent.increment_retry()
                continue

        return False

    def _are_blockers_resolved(self, blocked_on: List[str]) -> bool:
        """Check if all blocking tasks are complete."""
        for task_id in blocked_on:
            task = self.plan.get_task_by_id(task_id)
            if task and task.status != TaskStatus.COMPLETE:
                return False
        return True

    def _event_resolves_blockers(self, event: CIEvent, blocked_on: List[str]) -> bool:
        """Check if a CI event resolves our blockers."""
        # Build success or PR merge might resolve blockers
        if event.event_type in (
            CIEventType.BUILD_SUCCESS.value,
            CIEventType.PR_MERGED.value,
        ):
            # Re-check blockers after this event
            return self._are_blockers_resolved(blocked_on)
        return False

    async def _wait_for_pr_merge(self, pr_url: str) -> bool:
        """Wait for a PR to be merged."""
        # Extract PR number from URL (simplified)
        # In practice, you'd parse the URL properly
        try:
            pr_info = await self.ci_provider.wait_for_pr_merge(
                pr_number=1,  # Would need to extract from URL
                timeout=600,
            )
            return pr_info.is_merged()
        except (TimeoutError, RuntimeError):
            return False

    def _get_next_task(self, agent: AgentInstance) -> Optional[Task]:
        """Get the next available task for an agent."""
        available = self.plan.get_available_tasks_for_role(agent.role)
        return available[0] if available else None

    async def _on_ci_event(self, event: CIEvent):
        """Handle CI events."""
        await self._ci_event_queue.put(event)

    def stop(self):
        """Stop the loop."""
        self._running = False

File: src/lifecycle/loop.js
/**
 * @file Agent lifecycle loop - continuous retry with context reset.
 * @module lifecycle/loop
 */

import { readFile } from 'node:fs/promises';
import { LoopResultType } from '../types/index.js';
import { LifecycleState } from '../types/index.js';
import { getConfig } from '../config/index.js';
import { ContextBuilder } from './context.js';
import { CIEventType } from '../ci/events.js';

/**
 * Utility to sleep for a given time.
 * @param {number} ms - Milliseconds to sleep
 * @returns {Promise<void>}
 */
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Result of a loop iteration.
 */
export class LoopResult {
  /**
   * Create a LoopResult.
   * @param {Object} props - Properties
   * @param {string} props.resultType - Result type (LoopResultType value)
   * @param {string|null} [props.taskId=null] - Task ID if applicable
   * @param {string} [props.summary=''] - Result summary
   * @param {string[]} [props.blockedOn=[]] - Task IDs if blocked
   * @param {string|null} [props.prUrl=null] - PR URL if created
   * @param {string|null} [props.error=null] - Error message if any
   * @param {number} [props.spawnCount=0] - Number of spawns
   * @param {number} [props.retryCount=0] - Number of retries
   */
  constructor({
    resultType,
    taskId = null,
    summary = '',
    blockedOn = [],
    prUrl = null,
    error = null,
    spawnCount = 0,
    retryCount = 0,
  }) {
    /** @type {string} */
    this.resultType = resultType;
    /** @type {string|null} */
    this.taskId = taskId;
    /** @type {string} */
    this.summary = summary;
    /** @type {string[]} */
    this.blockedOn = blockedOn;
    /** @type {string|null} */
    this.prUrl = prUrl;
    /** @type {string|null} */
    this.error = error;
    /** @type {number} */
    this.spawnCount = spawnCount;
    /** @type {number} */
    this.retryCount = retryCount;
  }

  /**
   * Check if result indicates success.
   * @returns {boolean}
   */
  isSuccess() {
    return this.resultType === LoopResultType.TASK_COMPLETE || this.resultType === LoopResultType.PR_CREATED;
  }

  /**
   * Check if result indicates failure.
   * @returns {boolean}
   */
  isFailure() {
    return this.resultType === LoopResultType.MAX_RETRIES || this.resultType === LoopResultType.ERROR;
  }

  /**
   * Convert to plain object for JSON serialization.
   * @returns {import('../types/index.js').LoopResultData}
   */
  toDict() {
    return {
      resultType: this.resultType,
      taskId: this.taskId,
      summary: this.summary,
      blockedOn: this.blockedOn,
      prUrl: this.prUrl,
      error: this.error,
      spawnCount: this.spawnCount,
      retryCount: this.retryCount,
    };
  }

  /**
   * Create from plain object.
   * @param {import('../types/index.js').LoopResultData} data
   * @returns {LoopResult}
   */
  static fromDict(data) {
    return new LoopResult({
      resultType: data.resultType,
      taskId: data.taskId ?? null,
      summary: data.summary ?? '',
      blockedOn: data.blockedOn ?? [],
      prUrl: data.prUrl ?? null,
      error: data.error ?? null,
      spawnCount: data.spawnCount ?? 0,
      retryCount: data.retryCount ?? 0,
    });
  }
}

/**
 * The agent lifecycle loop - continuous retry mechanism with context reset.
 *
 * Key features:
 * - Agents work until they hit a natural breakpoint
 * - At breakpoints (task complete, blocked, PR created), agents are terminated
 * - Fresh agents are spawned with a context summary, preventing context rot
 * - CI events trigger immediate retry checks for blocked agents
 */
export class AgentLifecycleLoop {
  /**
   * Create an AgentLifecycleLoop.
   * @param {Object} props - Properties
   * @param {string} props.repoDir - Repository directory
   * @param {import('../plan/models.js').ProjectPlan} props.plan - Project plan
   * @param {import('../ci/interface.js').CIProvider} props.ciProvider - CI provider
   * @param {import('../runtime/process.js').TerminalManager} props.terminalManager - Terminal manager
   * @param {import('../runtime/branches.js').BranchManager} props.branchManager - Branch manager
   * @param {import('../runtime/workspace.js').WorkspaceManager} props.workspaceManager - Workspace manager
   * @param {string} props.commFilePath - Path to communications.json
   * @param {import('../personas/generator.js').ClaudeMdGenerator} props.claudeMdGenerator - Claude MD generator
   * @param {number} [props.maxRetries=100] - Maximum retries
   * @param {number} [props.retryInterval=30000] - Retry interval in ms
   * @param {string} [props.snapshotDir] - Directory for context snapshots
   */
  constructor({
    repoDir,
    plan,
    ciProvider,
    terminalManager,
    branchManager,
    workspaceManager,
    commFilePath,
    claudeMdGenerator,
    maxRetries = 100,
    retryInterval = 30000,
    snapshotDir,
  }) {
    /** @type {string} */
    this.repoDir = repoDir;
    /** @type {import('../plan/models.js').ProjectPlan} */
    this.plan = plan;
    /** @type {import('../ci/interface.js').CIProvider} */
    this.ciProvider = ciProvider;
    /** @type {import('../runtime/process.js').TerminalManager} */
    this.terminalManager = terminalManager;
    /** @type {import('../runtime/branches.js').BranchManager} */
    this.branchManager = branchManager;
    /** @type {import('../runtime/workspace.js').WorkspaceManager} */
    this.workspaceManager = workspaceManager;
    /** @type {string} */
    this.commFilePath = commFilePath;
    /** @type {import('../personas/generator.js').ClaudeMdGenerator} */
    this.claudeMdGenerator = claudeMdGenerator;
    /** @type {number} */
    this.maxRetries = maxRetries;
    /** @type {number} */
    this.retryInterval = retryInterval;
    /** @type {string} */
    this.snapshotDir = snapshotDir ?? `${repoDir}/.state/snapshots`;

    /** @type {ContextBuilder} */
    this.contextBuilder = new ContextBuilder(repoDir);

    /** @private @type {boolean} */
    this._running = false;
    /** @private @type {import('../ci/interface.js').CIEvent[]} */
    this._ciEventQueue = [];
  }

  /**
   * Run the lifecycle loop for an agent.
   *
   * @param {import('../personas/models.js').AgentInstance} agent - Agent instance
   * @param {import('../plan/models.js').Task} task - Initial task
   * @returns {Promise<LoopResult>}
   */
  async runAgentLoop(agent, task) {
    this._running = true;
    let currentTask = task;

    // Subscribe to CI events
    await this.ciProvider.subscribe(this._onCIEvent.bind(this));

    try {
      while (this._running && agent.retryCount < this.maxRetries) {
        // 1. Capture snapshot
        const snapshot = await this.contextBuilder.captureSnapshot(
          agent,
          currentTask,
          this.commFilePath
        );
        await this.contextBuilder.saveSnapshot(snapshot, this.snapshotDir);

        // 2. Build context summary
        const contextSummary = this.contextBuilder.buildContextSummary(snapshot);

        // 3. Spawn fresh agent with context
        const agentProcess = await this._spawnFreshAgent(agent, currentTask, contextSummary);
        agent.incrementSpawn();

        // 4. Wait for breakpoint
        const result = await this._waitForBreakpoint(agent, agentProcess);

        // 5. Handle breakpoint
        if (result.resultType === LoopResultType.TASK_COMPLETE) {
          console.log(`[Loop] Agent ${agent.agentId} completed task ${result.taskId}`);

          const nextTask = this._getNextTask(agent);
          if (nextTask) {
            currentTask = nextTask;
            agent.setWorking(currentTask);
            continue;
          } else {
            agent.setComplete();
            return result;
          }
        } else if (result.resultType === LoopResultType.BLOCKED) {
          console.log(`[Loop] Agent ${agent.agentId} blocked on ${result.blockedOn.join(', ')}`);
          agent.setBlocked(result.blockedOn, result.summary);

          // Terminate agent
          await this.terminalManager.terminate(agent.agentId);

          // Wait for unblock
          const unblocked = await this._waitForUnblock(agent, result.blockedOn);
          if (!unblocked) {
            return result;
          }

          agent.setIdle();
          continue;
        } else if (result.resultType === LoopResultType.PR_CREATED) {
          console.log(`[Loop] Agent ${agent.agentId} created PR: ${result.prUrl}`);
          agent.setPRPending(result.prUrl, result.taskId);

          // Terminate agent
          await this.terminalManager.terminate(agent.agentId);

          // Wait for PR merge
          const merged = await this._waitForPRMerge(result.prUrl);
          if (!merged) {
            return result;
          }

          const nextTask = this._getNextTask(agent);
          if (nextTask) {
            currentTask = nextTask;
            agent.setWorking(currentTask);
            continue;
          } else {
            agent.setComplete();
            return result;
          }
        } else if (result.resultType === LoopResultType.ERROR) {
          console.error(`[Loop] Agent ${agent.agentId} error: ${result.error}`);
          agent.incrementRetry();
          await this.terminalManager.terminate(agent.agentId);
          continue;
        } else {
          return result;
        }
      }

      // Max retries exceeded
      return new LoopResult({
        resultType: LoopResultType.MAX_RETRIES,
        taskId: currentTask?.id ?? null,
        spawnCount: agent.spawnCount,
        retryCount: agent.retryCount,
      });
    } finally {
      await this.ciProvider.unsubscribe(this._onCIEvent.bind(this));
      await this.terminalManager.terminate(agent.agentId);
    }
  }

  /**
   * Spawn a fresh agent with context summary.
   * @private
   * @param {import('../personas/models.js').AgentInstance} agent
   * @param {import('../plan/models.js').Task} task
   * @param {string} contextSummary
   * @returns {Promise<import('../runtime/process.js').AgentProcess>}
   */
  async _spawnFreshAgent(agent, task, contextSummary) {
    // Generate .claude.md with context
    const claudeMdContent = this.claudeMdGenerator.generate(
      agent.personaConfig,
      task,
      agent.branch,
      contextSummary
    );

    // Inject into workspace
    let workspacePath = this.workspaceManager.getSandbox(agent.agentId);
    if (!workspacePath) {
      workspacePath = await this.workspaceManager.createSandbox(agent.agentId);
    }

    await this.workspaceManager.injectClaudeMd(agent.agentId, claudeMdContent);

    // Build initial prompt
    const prompt = this._buildAgentPrompt(agent, task);

    // Spawn the agent
    return this.terminalManager.spawnClaudeAgent({
      agentId: agent.agentId,
      prompt,
      workingDir: workspacePath,
    });
  }

  /**
   * Build the initial prompt for a Claude agent.
   * @private
   * @param {import('../personas/models.js').AgentInstance} agent
   * @param {import('../plan/models.js').Task} task
   * @returns {string}
   */
  _buildAgentPrompt(agent, task) {
    return `You are agent "${agent.agentId}" (${agent.role}).

Your current task is: ${task.id} - ${task.description}

IMPORTANT: Read your .claude.md file for full context and instructions.

When you reach a natural stopping point (task complete, blocked, or PR created),
update communications.json with your breakpoint status.

Start by reading .claude.md and the current state of communications.json.`;
  }

  /**
   * Wait for an agent to reach a breakpoint.
   * @private
   * @param {import('../personas/models.js').AgentInstance} agent
   * @param {import('../runtime/process.js').AgentProcess} process
   * @returns {Promise<LoopResult>}
   */
  async _waitForBreakpoint(agent, process) {
    const config = getConfig();
    const checkInterval = config.breakpointCheckInterval;

    while (this._running && process.isRunning) {
      // Check communications.json for breakpoint
      const breakpoint = await this._checkForBreakpoint(agent.agentId);
      if (breakpoint) {
        return this._breakpointToResult(breakpoint);
      }

      await sleep(checkInterval);
    }

    // Process ended without breakpoint
    if (!this._running) {
      return new LoopResult({ resultType: LoopResultType.SHUTDOWN });
    }

    // Check one more time for breakpoint
    const breakpoint = await this._checkForBreakpoint(agent.agentId);
    if (breakpoint) {
      return this._breakpointToResult(breakpoint);
    }

    // Process ended without proper breakpoint - treat as error
    return new LoopResult({
      resultType: LoopResultType.ERROR,
      error: 'Agent process ended without signaling breakpoint',
    });
  }

  /**
   * Check communications.json for a breakpoint from an agent.
   * @private
   * @param {string} agentId
   * @returns {Promise<Object|null>}
   */
  async _checkForBreakpoint(agentId) {
    try {
      const content = await readFile(this.commFilePath, 'utf-8');
      const data = JSON.parse(content);
      const agentData = data[agentId] ?? {};

      const lifecycleState = agentData.lifecycleState ?? agentData.lifecycle_state ?? '';
      const breakpoint = agentData.breakpoint;

      if (['complete', 'blocked', 'pr_pending'].includes(lifecycleState) && breakpoint) {
        return breakpoint;
      }

      return null;
    } catch {
      return null;
    }
  }

  /**
   * Convert a breakpoint dict to a LoopResult.
   * @private
   * @param {Object} breakpoint
   * @returns {LoopResult}
   */
  _breakpointToResult(breakpoint) {
    const bpType = breakpoint.type ?? '';

    if (bpType === 'task_complete') {
      return new LoopResult({
        resultType: LoopResultType.TASK_COMPLETE,
        taskId: breakpoint.taskId ?? breakpoint.task_id ?? null,
        summary: breakpoint.summary ?? '',
      });
    } else if (bpType === 'blocked') {
      return new LoopResult({
        resultType: LoopResultType.BLOCKED,
        blockedOn: breakpoint.blockedOn ?? breakpoint.blocked_on ?? [],
        summary: breakpoint.reason ?? '',
      });
    } else if (bpType === 'pr_created') {
      return new LoopResult({
        resultType: LoopResultType.PR_CREATED,
        prUrl: breakpoint.prUrl ?? breakpoint.pr_url ?? null,
        taskId: breakpoint.taskId ?? breakpoint.task_id ?? null,
      });
    } else {
      return new LoopResult({
        resultType: LoopResultType.ERROR,
        error: `Unknown breakpoint type: ${bpType}`,
      });
    }
  }

  /**
   * Wait for blocking dependencies to be resolved.
   * @private
   * @param {import('../personas/models.js').AgentInstance} agent
   * @param {string[]} blockedOn
   * @returns {Promise<boolean>}
   */
  async _waitForUnblock(agent, blockedOn) {
    while (this._running && agent.retryCount < this.maxRetries) {
      // Check if all blockers are resolved
      if (this._areBlockersResolved(blockedOn)) {
        return true;
      }

      // Wait for CI event or timeout
      const event = await this._waitForCIEvent(this.retryInterval);

      if (event && this._eventResolvesBlockers(event, blockedOn)) {
        return true;
      }

      agent.incrementRetry();
    }

    return false;
  }

  /**
   * Check if all blocking tasks are complete.
   * @private
   * @param {string[]} blockedOn
   * @returns {boolean}
   */
  _areBlockersResolved(blockedOn) {
    const { TaskStatus } = require('../types/index.js');

    for (const taskId of blockedOn) {
      const task = this.plan.getTaskById(taskId);
      if (task && task.status !== TaskStatus.COMPLETE) {
        return false;
      }
    }
    return true;
  }

  /**
   * Check if a CI event resolves our blockers.
   * @private
   * @param {import('../ci/interface.js').CIEvent} event
   * @param {string[]} blockedOn
   * @returns {boolean}
   */
  _eventResolvesBlockers(event, blockedOn) {
    if (
      event.eventType === CIEventType.BUILD_SUCCESS ||
      event.eventType === CIEventType.PR_MERGED
    ) {
      return this._areBlockersResolved(blockedOn);
    }
    return false;
  }

  /**
   * Wait for a CI event with timeout.
   * @private
   * @param {number} timeoutMs
   * @returns {Promise<import('../ci/interface.js').CIEvent|null>}
   */
  async _waitForCIEvent(timeoutMs) {
    const startTime = Date.now();

    while (Date.now() - startTime < timeoutMs) {
      if (this._ciEventQueue.length > 0) {
        return this._ciEventQueue.shift();
      }
      await sleep(100);
    }

    return null;
  }

  /**
   * Extract PR number from a GitHub PR URL.
   * @private
   * @param {string} prUrl
   * @returns {number|null}
   */
  _extractPRNumber(prUrl) {
    const match = prUrl.match(/\/pull\/(\d+)/);
    if (match) {
      return parseInt(match[1], 10);
    }
    return null;
  }

  /**
   * Wait for a PR to be merged.
   * @private
   * @param {string} prUrl
   * @returns {Promise<boolean>}
   */
  async _waitForPRMerge(prUrl) {
    const prNumber = this._extractPRNumber(prUrl);
    if (prNumber === null) {
      console.error(`[Loop] Could not extract PR number from URL: ${prUrl}`);
      return false;
    }

    const config = getConfig();
    try {
      const prInfo = await this.ciProvider.waitForPRMerge(prNumber, config.prMergeTimeout);
      return prInfo.isMerged();
    } catch (error) {
      console.error(`[Loop] Error waiting for PR merge: ${error.message}`);
      return false;
    }
  }

  /**
   * Get the next available task for an agent.
   * @private
   * @param {import('../personas/models.js').AgentInstance} agent
   * @returns {import('../plan/models.js').Task|null}
   */
  _getNextTask(agent) {
    const available = this.plan.getAvailableTasksForRole(agent.role);
    return available.length > 0 ? available[0] : null;
  }

  /**
   * Handle CI events.
   * @private
   * @param {import('../ci/interface.js').CIEvent} event
   * @returns {void}
   */
  _onCIEvent(event) {
    this._ciEventQueue.push(event);
  }

  /**
   * Stop the loop.
   * @returns {void}
   */
  stop() {
    this._running = false;
  }
}

// Backwards compatibility alias
export const RalphWiggumLoop = AgentLifecycleLoop;
